Adding a simple default DNS query resolver

jabolina opened the pull request:
OK, the PR is as simple as it could be. It adds a service that receives a string and returns the list of addresses. This works using something in `/etc/hosts` and deploying a headless service in Kubernetes. But this is to start a discussion.

When running in Kubernetes, we might have multiple replicas, and the service will return the addresses of all pods:

```text
$ k get pods -n gingersnap
NAME                             READY   STATUS    RESTARTS   AGE
cache-manager-5869f7b9cb-542z5   1/1     Running   0          27s
cache-manager-5869f7b9cb-p2qcg   1/1     Running   0          27s
cache-manager-5869f7b9cb-z2jwv   1/1     Running   0          27s
db-syncer-6d6cf998d8-gm7cn       1/1     Running   0          27s
mysql-568447df7b-j4t6m           1/1     Running   0          27s
```

```text
$ k logs db-syncer-6d6cf998d8-gm7cn -n gingersnap
INFO exec  java -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -XX:+ExitOnOutOfMemoryError -cp "." -jar /deployments/db-syncer.jar
2023-01-19 13:25:01,980 INFO  [io.gin.cdc.ManagedEngine] (main) Starting service
FOUND ADDR => cache-manager/172.17.0.6
FOUND ADDR => cache-manager/172.17.0.9
FOUND ADDR => cache-manager/172.17.0.5
2023-01-19 13:25:02,003 INFO  [io.gin.k8s.CacheRuleInformer] (main) Kubernetes client not found, not watching config map
2023-01-19 13:25:02,055 INFO  [io.quarkus] (main) db-syncer 0.1-SNAPSHOT on JVM (powered by Quarkus 2.14.3.Final) started in 0.685s. Listening on: http://0.0.0.0:8080
2023-01-19 13:25:02,055 INFO  [io.quarkus] (main) Profile prod activated.
2023-01-19 13:25:02,055 INFO  [io.quarkus] (main) Installed features: [cdi, kubernetes-client, micrometer, resteasy-reactive, smallrye-context-propagation, smallrye-health, vertex]
```

We now have multiple `RemoteCacheManager`, one for each pod. Keeping in mind that we want to add dynamic membership, some pods can join and leave at any time. We have an issue with how to maintain all these caches.

If pods can arrive at any time, we have to create a new Debezium engine to start consuming events from the very start, otherwise, the pod might miss some data. As a side-effect, each pod might have a different offset and, therefore, different contents in the cache.
 
The approach I have in mind is similar to how we handle multiple rules. Each pod can fail independently, meaning that backoff and error handling is independent, too. And each pod would be able to proceed at its own pace. No tying all pods from a rule together.

I would like to get feedback before starting a change like that. [This](https://github.com/jabolina/gingersnap-k8s) is the setup I was running in Kubernetes.

ryanemerson commented:
> The approach I have in mind is similar to how we handle multiple rules. Each pod can fail independently, meaning that backoff and error handling is independent, too. And each pod would be able to proceed at its own pace. No tying all pods from a rule together.

So if we have `x` pods and `y` rules, does that mean we'll have `x*y` engines? I don't have an idea of how heavyweight these processes are, what kind of overhead will this have in terms of `db-syncer` resources and load on the DB?

jabolina commented:
Yeah, that would cause an explosion in the number of engines :disappointed: and I am also not confident enough to say that that would scale well for any DB provider.

One thing, maybe riskier, is to join an engine into another once they reach the same offset. But I'll need to investigate if that is feasible. Once an engine joins onto another, we treat the multiple caches as a single block.

---

I'll update the PR to add Ryan's suggestion of polling using the DNS resolver to identify joiners/leavers.

wburns commented:
Why would it have to be x*y? Would it not just be x for each pod? The rules already defined should be able to be lumped together. The problem would be rules coming up or down. But I _think_ we can just stop the engine(s) and restart them with the new rules configured (the offsets should be maintained).

ryanemerson commented on src/main/java/io/gingersnapproject/service/impl/DefaultDNSResolver.java:
```diff
@@ -0,0 +1,29 @@
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class DefaultDNSResolver implements DNSResolverService {
```
Do we need this to be it's own bean or can we just add the `resolve` code to `MembershipListener`?

jabolina replied:
We can add the method. I don't think we'll need to resolve a DNS from somewhere else.

wburns commented:
One thing to note about the DNS resolver here is we will want to have a configuration to disable this or only attempt the first address. For a cache-manager running as a headless k8s service we will want to connect to all services. But in the case of a bare metal clustered installation with a DNS record we don't want to connect to more than one.

In the log term we probably should check the client topology to see if it is clustered or not to determine if we should connect to all DNS records. But for the first pass we can just have a configuration option to enable or disable it.

jabolina commented:
@wburns Updated and stretched this a bit. Added the property `gingersnap.dynamic-membership` defaults to `false`. Added an interface that will deal with the dynamic member, the interface receives the URI to connect to. There is no implementation yet.

jabolina commented on src/main/java/io/gingersnapproject/cdc/configuration/Configuration.java:
```diff
@@ -3,11 +3,15 @@
 public interface Configuration {
 
+   @WithDefault("false")
+   boolean dynamicMembership();
```
looks like every property that starts with `gingersnap` must be mapped, or it throws an exception in the start-up.

ryanemerson commented on src/main/java/io/gingersnapproject/service/MembershipListener.java:
```diff
@@ -0,0 +1,131 @@
+      if (task != null || dmm.isUnsatisfied()) return;
+
+      if (dynamicMembership) {
+         task = new ReschedulingTask<>(executor, this::pollDNSInformation, __ -> false, 15, TimeUnit.SECONDS, __ -> true);
```
We should probably make this more frequent. Maybe Poll every second?

jabolina replied:
I think it doesn't change because the DNS records usually are cached. Running with minikube, IIRC takes around 20~30 secs to identify changes.

wburns replied:
I worry about loss of errors here as we don't log if a DNS error is thrown. Maybe we should log it as DEBUG at least?

wburns commented on src/main/java/io/gingersnapproject/service/MembershipListener.java:
```diff
@@ -0,0 +1,131 @@
+      return CompletableFutures.completedNull();
+   }
+
+   public static List<InetAddress> resolve(String query) {
```
Why is this static? It seems to be complicating the test code a lot. Do we expect someone to call this method outside of this class without the bean? I would expect to just inject the MembershipListener bean if I needed to use it.

jabolina replied:
We can leave it as a class method. I'll make it package-private, or we'll need to mock the `InetAddress.getAllByName` call.

wburns replied:
Sure, np

wburns commented on src/main/java/io/gingersnapproject/service/MembershipListener.java:
```diff
@@ -0,0 +1,131 @@
+   @Inject Configuration configuration;
+
+   // TODO: remove instance wrapper once implementation available.
+   @Inject Instance<DynamicMembershipManagement> dmm;
```
Do we expect possibly multiple implementations to listen to membership changes? Should this not just be part of the NotificationManager with a membership change event or something?

jabolina replied:
Besides the `ManagedEngine`, I see only something for metrics. I'll make an event for join and leave.

wburns commented on src/test/java/io/gingersnapproject/service/MembershipListenerTest.java:
```diff
@@ -0,0 +1,86 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Did you mean to have multiple test methods?

jabolina replied:
Only this one should cover everything. I tried an additional one calling `onStartup`, but it has to wait until the task is scheduled.

wburns replied:
Okay, was just wondering because this annotation is worthless right now otherwise.

wburns commented on src/test/java/io/gingersnapproject/service/MembershipListenerTest.java:
```diff
@@ -0,0 +1,79 @@
+      when(listener.resolve("cache-manager")).thenReturn(addresses);
+      listener.pollDNSInformation();
+      Assertions.assertEquals(new HashSet<>(addresses), Utils.extractField(MembershipListener.class, "members", listener));
+      verify(notificationMock, times(1)).memberLeaveEvent(any());
```
Does an eq verification not work for InetAddress here ?

jabolina replied:
Updated.

wburns commented on src/main/java/io/gingersnapproject/service/MembershipListener.java:
```diff
@@ -0,0 +1,131 @@
+public class MembershipListener {
+
+   private static final Logger log = LoggerFactory.getLogger(MembershipListener.class);
+   private static final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(r -> new Thread(r,"membership-listener"));
```
We should probably consolidate this with the other scheduler at some point.

wburns reviewed (approved):
LGTM, just really have one real question about test verification.

wburns commented:
Integrated into main, thanks @jabolina !
